enhancements/declarative-index-config.md

```diff
@@ -0,0 +1,426 @@
+	// TODO: regex version ranges
+	version: string
+    #serialized: json.Marshal({"pkg": packageName, "version": version})
+}
+
+#LabelConstraint: #Constraint & {
+	type:  "label"
+	value: string
+    #serialized: value
+}
+
+#GVKConstraint: #Constraint & {
+	type: "olm.gvk"
+	// TODO: regexes
+	group:   string
+	version: string
+	kind:    string
+    #serialized: json.Marshal({"group": group, "version": version, "kind": kind})
+}
+
+#Provides: {
+    type: string
+    #serialized: string
+    ...
+}
+
+#PackageProvides: #Provides & {
+    type:        "olm.package"
+	packageName: string
+	// TODO: regex semver version
+	version: string
+    #serialized: json.Marshal({"pkg": packageName, "version": version})
+}
+
+#GVKProvides: #Provides & {
+    type: "olm.gvk"
+	// TODO: regexes
+	group:   string
+	version: string
+	kind:    string
+    #serialized: json.Marshal({"group": group, "version": version, "kind": kind})
+}
+
+#LabelProvides: #Provides & {
+	type:  "label"
+	value: string
+    #serialized: value
+}
+
+#Node: {
+	bundlePath:   string
+	provides?:    [...#Provides]
+	constraints?: [...#ConstraintSet]
+}
+
+#Edge: {
+    from: string,
+    to: string,
+    _bad: *false | true
+}
+
+#EdgeMap: [Name=string]: {
+    [To=string]: #Edge & {
+        from: Name
+        to: To
+    }
+}
+
+// edge is a list of graph edges between nodes
+_edge: #EdgeMap
+
+// nedge is a list of rejected graph edges that should not be followed
+_nedge: #EdgeMap
+
+graph: nodes: [...#Node]
+```
+
+This definition will live in the [API repo](https://github.com/operator-framework/api) as `definition.CUE`. `opm` will import the CUE project and use the relevant portions for index representation and validation.
+
+#### Representation
+
+A new opm sub-command, `represent` will be introduced that will output a yaml/json representation of an index, i.e, `opm represent --output=json docker.io/my-namespace/my-index` will output the following json representation of the index:
+
+```output.json
+{
+    "graph": {
+        "nodes": [
+            {
+                "bundlePath": "quay.io/etcd",
+                "provides": [
+                    {
+                        "type": "olm.package",
+                        "version": "1.0.0",
+                        "packageName": "etcd"
+                    },
+                    {
+                        "type": "olm.gvk",
+                        "version": "v1beta2",
+                        "group": "etcd.database.coreos.com",
+                        "kind": "EtcdCluster"
+                    },
+                    {
+                        "type": "label",
+                        "value": "alpha"
+                    }
+                ],
+                "constraints": [
+                    [
+                        {
+                            "type": "label",
+                            "value": "LTS"
+                        }
+                    ]
+                ]
+            },
+        ]
+    }
+}
```

Comment:
can we see an example that has multiple versions (and even better, multiple channels) for a given package, with different upgrade edges (skips, skipRange, replaces)?

Reply:
@bparees updated the enhancement with packages containing channels(with their defined upgrade modes) instead of packages containing bundles which have channel information in them. With that changes, does [this](https://github.com/operator-framework/enhancements/pull/37/files#diff-1edbfba9fdddd2b74b5cd32c6de146a4f62e1400a93bee78aed2cbf6ebaaa1d7R110-R163) example look better?